agent/consul/state/stream_publisher.go

```diff
@@ -0,0 +1,439 @@
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+type EventPublisher struct {
+	store *Store
+
+	// topicBufferSize controls how many trailing events we keep in memory for
+	// each topic to avoid needing to snapshot again for re-connecting clients
+	// that may have missed some events. It may be zero for no buffering (the most
+	// recent event is always kept though). TODO
+	topicBufferSize int
+
+	// snapCacheTTL controls how long we keep snapshots in our cache before
+	// allowing them to be garbage collected and a new one made for subsequent
+	// requests for that topic and key. In general this should be pretty short to
+	// keep memory overhead of duplicated event data low - snapshots are typically
+	// not that expensive, but having a cache for a few seconds can help
+	// de-duplicate building the same snapshot over and over again when a
+	// thundering herd of watchers all subscribe to the same topic within a few
+	// seconds. TODO
+	snapCacheTTL time.Duration
+
+	// This lock protects the topicBuffers, snapCache and subsByToken maps.
+	lock sync.RWMutex
+
+	// topicBuffers stores the head of the linked-list buffer to publish events to
+	// for a topic.
+	topicBuffers map[agentpb.Topic]*stream.EventBuffer
+
+	// snapCache stores the head of any snapshot buffers still in cache if caching
+	// is enabled.
+	snapCache map[agentpb.Topic]map[string]*stream.EventSnapshot
+
+	// snapFns is the set of snapshot functions that were registered bound to the
+	// state store.
+	snapFns map[agentpb.Topic]stream.SnapFn
+
+	// subsByToken stores a list of Subscription objects outstanding indexed by a
+	// hash of the ACL token they used to subscribe so we can reload them if their
+	// ACL permissions change.
+	subsByToken map[string]map[*agentpb.SubscribeRequest]*stream.Subscription
+
+	// commitCh decouples the Commit call in the FSM hot path from distributing
+	// the resulting events.
+	commitCh chan commitUpdate
+
+	// staged stores the events to be published when the transaction is committed.
+	staged     []agentpb.Event
+	stagedLock sync.RWMutex
+}
+
+type commitUpdate struct {
+	tx     *txnWrapper
+	events []agentpb.Event
+}
+
+func NewEventPublisher(store *Store, topicBufferSize int, snapCacheTTL time.Duration) *EventPublisher {
+	e := &EventPublisher{
+		store:           store,
+		topicBufferSize: topicBufferSize,
+		snapCacheTTL:    snapCacheTTL,
+		topicBuffers:    make(map[agentpb.Topic]*stream.EventBuffer),
+		snapCache:       make(map[agentpb.Topic]map[string]*stream.EventSnapshot),
+		snapFns:         make(map[agentpb.Topic]stream.SnapFn),
+		subsByToken:     make(map[string]map[*agentpb.SubscribeRequest]*stream.Subscription),
+		commitCh:        make(chan commitUpdate, 64),
+	}
+
+	// create a local handler table
+	for topic, handlers := range topicRegistry {
+		fnCopy := handlers.Snapshot
+		e.snapFns[topic] = func(req *agentpb.SubscribeRequest, buf *stream.EventBuffer) (uint64, error) {
+			return fnCopy(e.store, req, buf)
+		}
+	}
```

Comment:
Is this done because the `topicRegistry` contains unbound functions, or is there another reason?

I was hoping to change this around and have `NewEventPublisher` accept a registry with bound functions.

Reply:
Yeah they are unbound as it makes it cleaner to use state store methods this way rather than stand-alone functions that need to be passed a ton of state. I don't feel very strongly, this is just the pattern we use elsewhere in the code base for RPC and HTTP handlers so I went with it.

Happy to change although I think the main thing to optimize for here is decoupling and easy maintenance and testing of the event handlers. Did you have a specific concern with this approach?

Reply:
I've been thinking about the other places in the code base where we do this as well. [This diff](https://github.com/hashicorp/consul/compare/dnephin/http-api-separate-package) was an experiment I did to remove the pattern from the http handlers, and it appears to work.

> Happy to change although I think the main thing to optimize for here is decoupling and easy maintenance and testing of the event handlers.

I very much agree with this. I think the unbound method registry has a number of disadvantages, and no advantages I could find.

Declaring the registry statically (in init) makes it more difficult to test the generic parts that read the registry. If `New` accepts a map a test can pass it a set of stubs so the generic part can easily be tested independently of any of the specific implementations.

Requiring an unbound method forces us to implement every function on the same type. I think that leads to sprawling structs and packages. A registry of functions allows us to use functions from other packages.

We can keep the function signature small by passing in bound methods for the functions, which meets the goal of not passing around a ton of state.